rest/route/host_agent.go

```diff
@@ -778,28 +683,77 @@ func assignNextAvailableTask(ctx context.Context, taskQueue *model.TaskQueue, di
-		})
 
-		if !ok {
+		if !dispatchedTask {
 			continue
 		}
 
 		return nextTask, false, nil
 	}
+
 	return nil, false, nil
 }
 
+// checkHostTaskGroupAfterDispatch checks that the task group max hosts is
+// still respected after the task has already been assigned to the host and that
+// for single-host task groups, it is dispatching the next task in the group.
+//
+// It's possible for dispatchers on different app servers to race, assigning
+// different tasks in a task group to more hosts than the task group's max
+// hosts. Therefore, we must check that the number of hosts running this task
+// group does not exceed the max after assigning the task to the host. If it
+// exceeds the max host limit, this will return true to indicate that the host
+// should not run this task.
+func checkHostTaskGroupAfterDispatch(h *host.Host, t *task.Task) (isDispatchable bool, err error) {
+	var minTaskGroupOrderNum int
+	if t.TaskGroupMaxHosts == 1 {
+		// Regardless of how many hosts are running tasks, if this host is
+		// running the earliest task in the task group we should continue.
+		minTaskGroupOrderNum, err = host.MinTaskGroupOrderRunningByTaskSpec(t.TaskGroup, t.BuildVariant, t.Project, t.Version)
+		if err != nil {
+			return false, errors.Wrap(err, "getting min task group order")
+		}
+		// minTaskGroupOrderNum is only available if a host has that information
+		// cached, which was not always the case. If some host doesn't have the
+		// task group order cached, then minTaskGroupOrderNum will be 0. For
+		// backward compatibility in case it is 0, this will later fall back to
+		// counting the hosts to check that max hosts is respected.
+		if minTaskGroupOrderNum != 0 && minTaskGroupOrderNum < t.TaskGroupOrder {
+			return false, errors.Errorf("current task is order %d but another host is running %d", t.TaskGroupOrder, minTaskGroupOrderNum)
+		}
+		if t.TaskGroupOrder > 1 {
+			// If the previous task in the single-host task group has yet to run
+			// and should run, then wait for the previous task to run.
+			tgTasks, err := task.FindTaskGroupFromBuild(t.BuildId, t.TaskGroup)
+			if err != nil {
+				return false, errors.Wrap(err, "finding task group from build")
+			}
+			for _, tgTask := range tgTasks {
+				if tgTask.TaskGroupOrder == t.TaskGroupOrder {
+					break
+				}
+				if tgTask.TaskGroupOrder < t.TaskGroupOrder && tgTask.IsHostDispatchable() && !tgTask.Blocked() {
+					return false, errors.Errorf("an earlier task in the task group ('%s') is still dispatchable", tgTask.DisplayName)
+				}
+			}
+		}
+	}
+
+	// For multiple-host task groups and single-host task groups without order
+	// cached in the host, check that max hosts is respected.
+	if minTaskGroupOrderNum == 0 {
+		numHosts, err := host.NumHostsByTaskSpec(t.TaskGroup, t.BuildVariant, t.Project, t.Version)
+		if err != nil {
+			return false, errors.Wrap(err, "getting number of hosts running task group")
+		}
+		if numHosts > t.TaskGroupMaxHosts {
```

Comment:
Since we only run this function for hosts that are new to the group, wouldn't we want to check if `numHosts >= t.TaskGroupMaxHosts` since adding one more host would breach the limit?

Reply:
The max hosts check is done after the task is already dispatched because we can't be sure when doing the dispatch updates whether or not max hosts is exceeded yet.

Reply:
Ah I see, thanks for clarifying